lib/tree.rb

```diff
@@ -1,64 +1,165 @@
+    list << {key: key, value: value}
+    return list
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: o(n)
+  def add_helper(current, key, value)
+    return TreeNode.new(key, value) if current.nil?
+    
+    if key <= current.key
+      current.left = add_helper(current.left, key, value)
+    else
+      current.right = add_helper(current.right, key, value)
+    end
+    return current
+  end
+  
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(1)
+  # Space Complexity: O1
   def find(key)
-    raise NotImplementedError
+    current = @root
+    until current.nil?
+      if current.key == key
+        return current.value
+      elsif key > current.key
+        current = current.right
+      else
+        current = current.left
+      end
+    end
+    return nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: On
+  # Space Complexity: On
+  def inorder_helper 
+    return inorder_helper(@root, [])
+    list = left.inorder(list) unless left.nil?
+    list << {key: key, value: value}
+    list = right.inorder(list) unless right.nil?
+    return list
+  end
+  
   def inorder
-    raise NotImplementedError
+    list = []
+    return list if @root.nil?
+    return @root.inorder(list)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: On
+  # Space Complexity: On
   def preorder
-    raise NotImplementedError
+    list = []
+    return list if @root.nil?
+    return @root.preorder([])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity:  On
+  # Space Complexity: On
   def postorder
-    raise NotImplementedError
+    list = []
+    return list if @root.nil?
+    return @root.postorder([])
   end
-
+  
   # Time Complexity: 
   # Space Complexity: 
+  def max (num1, num2)
+    if num1 > num2
+      return num1
+    else
+      return num2
+    end
+  end
+  
+  def height_helper(current)
+    return 0 if current.nil?
+    return 1 + max(height_helper(current.left), height_helper(current.right))
+  end
+  
   def height
-    raise NotImplementedError
+    return height_helper(@root)
+  end
+  
+  def delete_helper(current, key)
+    return nil if current.nil?
+    
+    if current.key > key
+      current.left = delete_helper(current.left, key)
+      return current
+    end
+    
+    if current.key < key
+      current.right = delete_helper(current.right, key)
+      return current
+    end
+    
+    if current.left.nil? && current.right.nil?
+      return nil
+    end
+    
+    if current.left.nil?
+      return current.right
+    end
+    
+    if current.right.nil?
+      return current.left
+    end
+    
+  end
+  
+  def delete(key)
+    value = self.find(key)
```

Comment:
Not sure what value is doing?

This is a good start at the `delete` method, well done.  It's not quite all the way.